provenance_lib/parse.py

```diff
@@ -0,0 +1,362 @@
+
+
+class Archive:
+    """
+    Lightly-processed contents of a single QIIME 2 Archive, an Archive
+    contains a non-hierarchical pool of unique ProvNodes
+    """
+
+    # TODO: Can we assume all Q2 archives contain only unique Results?
+    # If not, Archive should filter duplicates.
+
+    # TODO: add property for archive version?
+    # TODO: UUID class with basic validation?
+    # NOTE: static type declarations like this break testing unless the class
+    # is forward declared (ProvNode must be above Archive or NameError).
+    _number_of_results: int
+    _archive_contents: Dict[str, ProvNode]
+    _archive_md: _ResultMetadata
+
+    @property
+    def root_uuid(self):
+        return self._archive_md.uuid
+
+    @property
+    def archive_type(self):
+        return self._archive_md.type
+
+    @property
+    def archive_format(self):
+        return self._archive_md.format
+
+    def get_result(self, uuid):
+        return self._archive_contents[uuid]
+
+    def __str__(self):
+        return f"Archive(Root: {self.root_uuid})"
+
+    def __repr__(self):
+        r_str = (f"Archive(Root: {self.root_uuid}, Semantic Type: "
+                 f"{self.archive_type}, Format: {self.archive_format})\n"
+                 "Contains Results:\n")
+        for uuid in self._archive_contents.keys():
+            r_str += str(uuid)
+            r_str += "\n"
+        return r_str
+
+    def __init__(self, archive_fp: str):
+        self._archive_md: None
+        self._archive_contents: None
+        self._number_of_results = 0
+
+        # Get archive metadata including root uuid
+        with zipfile.ZipFile(archive_fp) as zf:
+            all_filenames = zf.namelist()
+            root_metadata_fps = filter(self._is_root_metadata_file,
+                                       all_filenames)
+
+            try:
+                root_metadata_fp = next(root_metadata_fps)
+            except StopIteration:
+                raise ValueError("Malformed Archive: "
+                                 "no top-level metadata.yaml file")
+
+            try:
+                next(root_metadata_fps)
+                raise ValueError("Malformed Archive: "
+                                 "multiple top-level metadata.yaml files")
+            except StopIteration:
+                pass
+
+            self._archive_md = _ResultMetadata(zf, root_metadata_fp)
+            self._populate_archive(zf)
+
+    def _populate_archive(self, zf: zipfile):
+        self._archive_contents = {}
+
+        prov_data_filenames = filter(self._is_prov_data, zf.namelist())
+        prov_data_fps = list(map(pathlib.Path, prov_data_filenames))
+
+        all_uuids = set()
+        for fp in prov_data_fps:
+            uuid = ""
+            if 'artifacts' not in fp.parts:
+                uuid = fp.parts[0]
+            else:
+                uuid = self._get_nonroot_uuid(fp)
+            all_uuids.add(uuid)
+
+        # make a provnode for each UUID
+        for uuid in all_uuids:
```

Comment:
Could this happen in the previous loop, as you're identifying the uuids?

Reply:
Addressed in edc2e9a337d1160b8c0afe688427b5387df3ed7f. This refactor also cuts a lot of redundant filepath filtering.